File: src/metering/usage_tracker.py
"""
Usage Tracker — ROIaaS Phase 4

Tracks CLI command usage per license key:
- Command invocations (cook, plan, fix, etc.)
- Agent calls (planner, researcher, fullstack-developer, etc.)
- Pipeline runs (PEV cycles)
- Free tier enforcement: 10 commands/day

Storage: SQLite (~/.mekong/raas/tenants.db)

Usage:
    from src.metering.usage_tracker import UsageTracker
    tracker = UsageTracker()
    tracker.track_command("license-123", "cook")
    usage = tracker.get_daily_usage("license-123")
"""

import sqlite3
import hashlib
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Dict, List, Optional, Any


# Free tier limits
FREE_TIER_LIMITS = {
    "commands_per_day": 10,
    "agents_per_day": 5,
    "pipelines_per_day": 3,
}


@dataclass
class DailyUsage:
    """Daily usage summary."""
    date: str  # YYYY-MM-DD
    total_commands: int = 0
    total_agents: int = 0
    total_pipelines: int = 0
    command_breakdown: Dict[str, int] = field(default_factory=dict)
    agent_breakdown: Dict[str, int] = field(default_factory=dict)


@dataclass
class UsageReport:
    """Multi-day usage report."""
    license_key_hash: str
    period_days: int
    total_commands: int
    total_agents: int
    total_pipelines: int
    daily_reports: List[DailyUsage] = field(default_factory=list)


class UsageTracker:
    """SQLite-backed usage tracker."""

    def __init__(self, db_path: Optional[Path] = None) -> None:
        """Initialize usage tracker.

        Args:
            db_path: Optional database path override for testing.
        """
        if db_path is None:
            db_path = Path.home() / ".mekong" / "raas" / "tenants.db"

        self._db_path = Path(db_path)
        self._db_path.parent.mkdir(parents=True, exist_ok=True)
        self._conn = sqlite3.connect(str(self._db_path), timeout=10)
        self._conn.execute("PRAGMA journal_mode=WAL")
        self._conn.row_factory = sqlite3.Row
        self._init_tables()

    def _init_tables(self) -> None:
        """Create usage_events table if not exists."""
        self._conn.executescript("""
            CREATE TABLE IF NOT EXISTS usage_events (
                id TEXT PRIMARY KEY,
                license_key_hash TEXT NOT NULL,
                event_type TEXT NOT NULL,
                event_name TEXT NOT NULL,
                units INTEGER NOT NULL DEFAULT 1,
                metadata TEXT,
                timestamp TEXT NOT NULL
            );

            CREATE INDEX IF NOT EXISTS idx_usage_license_date
            ON usage_events(license_key_hash, substr(timestamp, 1, 10));

            CREATE INDEX IF NOT EXISTS idx_usage_type
            ON usage_events(event_type);
        """)

    def _hash_license_key(self, license_key: str) -> str:
        """Hash license key for privacy."""
        return hashlib.sha256(license_key.encode()).hexdigest()

    def track_command(
        self,
        license_key: str,
        command: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Track a command execution."""
        self._track_event(
            license_key,
            "command",
            command,
            metadata=metadata,
        )

    def track_agent_call(
        self,
        license_key: str,
        agent_name: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Track an agent call."""
        self._track_event(
            license_key,
            "agent_call",
            agent_name,
            metadata=metadata,
        )

    def track_pipeline_run(
        self,
        license_key: str,
        pipeline_type: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Track a pipeline run."""
        self._track_event(
            license_key,
            "pipeline_run",
            pipeline_type,
            metadata=metadata,
        )

    def _track_event(
        self,
        license_key: str,
        event_type: str,
        event_name: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Internal: Track a usage event."""
        import uuid

        license_key_hash = self._hash_license_key(license_key)
        event_id = str(uuid.uuid4())
        timestamp = datetime.now(timezone.utc).isoformat()

        self._conn.execute(
            """
            INSERT INTO usage_events
            (id, license_key_hash, event_type, event_name, units, metadata, timestamp)
            VALUES (?, ?, ?, ?, ?, ?, ?)
            """,
            (
                event_id,
                license_key_hash,
                event_type,
                event_name,
                1,
                str(metadata or {}),
                timestamp,
            ),
        )
        self._conn.commit()

    def get_daily_usage(
        self,
        license_key: str,
        date: Optional[str] = None,
    ) -> DailyUsage:
        """Get daily usage for a license key."""
        license_key_hash = self._hash_license_key(license_key)
        target_date = date or datetime.now(timezone.utc).strftime("%Y-%m-%d")

        # Get commands
        cmd_rows = self._conn.execute(
            """
            SELECT event_name, SUM(units) as count
            FROM usage_events
            WHERE license_key_hash = ?
              AND event_type = 'command'
              AND substr(timestamp, 1, 10) = ?
            GROUP BY event_name
            """,
            (license_key_hash, target_date),
        ).fetchall()

        # Get agents
        agent_rows = self._conn.execute(
            """
            SELECT event_name, SUM(units) as count
            FROM usage_events
            WHERE license_key_hash = ?
              AND event_type = 'agent_call'
              AND substr(timestamp, 1, 10) = ?
            GROUP BY event_name
            """,
            (license_key_hash, target_date),
        ).fetchall()

        # Get pipelines
        pipeline_row = self._conn.execute(
            """
            SELECT SUM(units) as count
            FROM usage_events
            WHERE license_key_hash = ?
              AND event_type = 'pipeline_run'
              AND substr(timestamp, 1, 10) = ?
            """,
            (license_key_hash, target_date),
        ).fetchone()

        # Build result
        command_breakdown: Dict[str, int] = {}
        total_commands = 0
        for row in cmd_rows:
            command_breakdown[row["event_name"]] = row["count"] or 0
            total_commands += row["count"] or 0

        agent_breakdown: Dict[str, int] = {}
        total_agents = 0
        for row in agent_rows:
            agent_breakdown[row["event_name"]] = row["count"] or 0
            total_agents += row["count"] or 0

        return DailyUsage(
            date=target_date,
            total_commands=total_commands,
            total_agents=total_agents,
            total_pipelines=pipeline_row["count"] or 0 if pipeline_row else 0,
            command_breakdown=command_breakdown,
            agent_breakdown=agent_breakdown,
        )

    def is_free_tier_exceeded(
        self,
        license_key: str,
        date: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Check if free tier limit exceeded."""
        usage = self.get_daily_usage(license_key, date)

        if usage.total_commands >= FREE_TIER_LIMITS["commands_per_day"]:
            return {
                "exceeded": True,
                "reason": f"Command limit exceeded: {usage.total_commands}/{FREE_TIER_LIMITS['commands_per_day']}",
            }

        if usage.total_agents >= FREE_TIER_LIMITS["agents_per_day"]:
            return {
                "exceeded": True,
                "reason": f"Agent limit exceeded: {usage.total_agents}/{FREE_TIER_LIMITS['agents_per_day']}",
            }

        if usage.total_pipelines >= FREE_TIER_LIMITS["pipelines_per_day"]:
            return {
                "exceeded": True,
                "reason": f"Pipeline limit exceeded: {usage.total_pipelines}/{FREE_TIER_LIMITS['pipelines_per_day']}",
            }

        return {"exceeded": False}

    def get_free_tier_remaining(self, usage: DailyUsage) -> Dict[str, int]:
        """Get remaining free tier quota."""
        return {
            "commands_remaining": max(0, FREE_TIER_LIMITS["commands_per_day"] - usage.total_commands),
            "agents_remaining": max(0, FREE_TIER_LIMITS["agents_per_day"] - usage.total_agents),
            "pipelines_remaining": max(0, FREE_TIER_LIMITS["pipelines_per_day"] - usage.total_pipelines),
        }

    def get_usage_report(self, license_key: str, days: int = 7) -> UsageReport:
        """Get multi-day usage report."""
        from datetime import timedelta

        today = datetime.now(timezone.utc)
        daily_reports: List[DailyUsage] = []
        total_commands = 0
        total_agents = 0
        total_pipelines = 0

        for i in range(days):
            date = today - timedelta(days=i)
            date_str = date.strftime("%Y-%m-%d")
            daily_usage = self.get_daily_usage(license_key, date_str)
            daily_reports.append(daily_usage)
            total_commands += daily_usage.total_commands
            total_agents += daily_usage.total_agents
            total_pipelines += daily_usage.total_pipelines

        return UsageReport(
            license_key_hash=self._hash_license_key(license_key),
            period_days=days,
            total_commands=total_commands,
            total_agents=total_agents,
            total_pipelines=total_pipelines,
            daily_reports=daily_reports,
        )

    def close(self) -> None:
        """Close database connection."""
        self._conn.close()


# Singleton instance
_tracker: Optional[UsageTracker] = None


def get_tracker() -> UsageTracker:
    """Get global usage tracker instance."""
    global _tracker
    if _tracker is None:
        _tracker = UsageTracker()
    return _tracker

File: src/metering/usage-tracker.ts
/**
 * Usage Tracker — ROIaaS Phase 4
 *
 * Tracks CLI command usage per license key:
 * - Command invocations (cook, plan, fix, etc.)
 * - Agent calls (planner, researcher, fullstack-developer, etc.)
 * - Pipeline runs (PEV cycles)
 * - Free tier enforcement: 10 commands/day
 *
 * Storage: SQLite (~/.mekong/raas/tenants.db)
 *
 * Usage:
 * ```typescript
 * const tracker = UsageTracker.getInstance();
 * await tracker.trackCommand('license-123', 'cook', { complexity: 'standard' });
 * const usage = await tracker.getDailyUsage('license-123');
 * const remaining = tracker.getFreeTierRemaining(usage);
 * ```
 */

import { Database } from 'better-sqlite3';
import { join } from 'path';
import { homedir } from 'os';
import { mkdirSync, existsSync } from 'fs';
import * as crypto from 'crypto';

/**
 * Database path
 */
const DB_PATH = join(homedir(), '.mekong', 'raas', 'tenants.db');

/**
 * Free tier limits
 */
export const FREE_TIER_LIMITS = {
  commandsPerDay: 10,
  agentsPerDay: 5,
  pipelinesPerDay: 3,
};

/**
 * Billable event types
 */
export enum UsageEventType {
  COMMAND = 'command',
  AGENT_CALL = 'agent_call',
  PIPELINE_RUN = 'pipeline_run',
}

/**
 * Usage event structure
 */
export interface UsageEvent {
  id: string;
  licenseKeyHash: string;
  eventType: UsageEventType;
  eventName: string; // command name, agent name, pipeline type
  units: number;
  metadata?: Record<string, unknown>;
  timestamp: string; // ISO 8601
}

/**
 * Daily usage summary
 */
export interface DailyUsage {
  date: string; // YYYY-MM-DD
  totalCommands: number;
  totalAgents: number;
  totalPipelines: number;
  commandBreakdown: Record<string, number>;
  agentBreakdown: Record<string, number>;
}

/**
 * Usage tracker class
 */
export class UsageTracker {
  private static instance: UsageTracker;
  private db: Database;

  private constructor() {
    // Ensure directory exists
    const dbDir = join(homedir(), '.mekong', 'raas');
    if (!existsSync(dbDir)) {
      mkdirSync(dbDir, { recursive: true });
    }

    // Initialize SQLite
    this.db = new Database(DB_PATH);
    this.db.pragma('journal_mode = WAL');
    this.initTables();
  }

  /**
   * Get singleton instance
   */
  static getInstance(): UsageTracker {
    if (!UsageTracker.instance) {
      UsageTracker.instance = new UsageTracker();
    }
    return UsageTracker.instance;
  }

  /**
   * Initialize database tables
   */
  private initTables(): void {
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS usage_events (
        id TEXT PRIMARY KEY,
        license_key_hash TEXT NOT NULL,
        event_type TEXT NOT NULL,
        event_name TEXT NOT NULL,
        units INTEGER NOT NULL DEFAULT 1,
        metadata TEXT,
        timestamp TEXT NOT NULL
      );

      CREATE INDEX IF NOT EXISTS idx_usage_license_date
      ON usage_events(license_key_hash, substr(timestamp, 1, 10));

      CREATE INDEX IF NOT EXISTS idx_usage_type
      ON usage_events(event_type);
    `);
  }

  /**
   * Hash license key for privacy
   */
  private hashLicenseKey(licenseKey: string): string {
    return crypto.createHash('sha256').update(licenseKey).digest('hex');
  }

  /**
   * Track a command execution
   */
  async trackCommand(
    licenseKey: string,
    command: string,
    metadata?: Record<string, unknown>
  ): Promise<void> {
    const licenseKeyHash = this.hashLicenseKey(licenseKey);
    const id = crypto.randomUUID();
    const timestamp = new Date().toISOString();

    const stmt = this.db.prepare(`
      INSERT INTO usage_events (id, license_key_hash, event_type, event_name, units, metadata, timestamp)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    `);

    stmt.run(
      id,
      licenseKeyHash,
      UsageEventType.COMMAND,
      command,
      1,
      JSON.stringify(metadata || {}),
      timestamp
    );
  }

  /**
   * Track an agent call
   */
  async trackAgentCall(
    licenseKey: string,
    agentName: string,
    metadata?: Record<string, unknown>
  ): Promise<void> {
    const licenseKeyHash = this.hashLicenseKey(licenseKey);
    const id = crypto.randomUUID();
    const timestamp = new Date().toISOString();

    const stmt = this.db.prepare(`
      INSERT INTO usage_events (id, license_key_hash, event_type, event_name, units, metadata, timestamp)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    `);

    stmt.run(
      id,
      licenseKeyHash,
      UsageEventType.AGENT_CALL,
      agentName,
      1,
      JSON.stringify(metadata || {}),
      timestamp
    );
  }

  /**
   * Track a pipeline run
   */
  async trackPipelineRun(
    licenseKey: string,
    pipelineType: string,
    metadata?: Record<string, unknown>
  ): Promise<void> {
    const licenseKeyHash = this.hashLicenseKey(licenseKey);
    const id = crypto.randomUUID();
    const timestamp = new Date().toISOString();

    const stmt = this.db.prepare(`
      INSERT INTO usage_events (id, license_key_hash, event_type, event_name, units, metadata, timestamp)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    `);

    stmt.run(
      id,
      licenseKeyHash,
      UsageEventType.PIPELINE_RUN,
      pipelineType,
      1,
      JSON.stringify(metadata || {}),
      timestamp
    );
  }

  /**
   * Get daily usage for a license key
   */
  getDailyUsage(licenseKey: string, date?: string): DailyUsage {
    const licenseKeyHash = this.hashLicenseKey(licenseKey);
    const targetDate = date || new Date().toISOString().slice(0, 10);

    // Get commands
    const commandsQuery = this.db.prepare(`
      SELECT event_name, SUM(units) as count
      FROM usage_events
      WHERE license_key_hash = ?
        AND event_type = 'command'
        AND substr(timestamp, 1, 10) = ?
      GROUP BY event_name
    `);

    // Get agents
    const agentsQuery = this.db.prepare(`
      SELECT event_name, SUM(units) as count
      FROM usage_events
      WHERE license_key_hash = ?
        AND event_type = 'agent_call'
        AND substr(timestamp, 1, 10) = ?
      GROUP BY event_name
    `);

    // Get pipelines
    const pipelinesQuery = this.db.prepare(`
      SELECT SUM(units) as count
      FROM usage_events
      WHERE license_key_hash = ?
        AND event_type = 'pipeline_run'
        AND substr(timestamp, 1, 10) = ?
    `);

    const commandRows = commandsQuery.all(licenseKeyHash, targetDate) as Array<{ event_name: string; count: number }>;
    const agentRows = agentsQuery.all(licenseKeyHash, targetDate) as Array<{ event_name: string; count: number }>;
    const pipelineRow = pipelinesQuery.get(licenseKeyHash, targetDate) as { count: number } | undefined;

    const commandBreakdown: Record<string, number> = {};
    let totalCommands = 0;
    for (const row of commandRows) {
      commandBreakdown[row.event_name] = row.count;
      totalCommands += row.count;
    }

    const agentBreakdown: Record<string, number> = {};
    let totalAgents = 0;
    for (const row of agentRows) {
      agentBreakdown[row.event_name] = row.count;
      totalAgents += row.count;
    }

    return {
      date: targetDate,
      totalCommands,
      totalAgents,
      totalPipelines: pipelineRow?.count || 0,
      commandBreakdown,
      agentBreakdown,
    };
  }

  /**
   * Check if free tier limit exceeded
   */
  isFreeTierExceeded(licenseKey: string, date?: string): {
    exceeded: boolean;
    reason?: string;
  } {
    const usage = this.getDailyUsage(licenseKey, date);

    if (usage.totalCommands >= FREE_TIER_LIMITS.commandsPerDay) {
      return {
        exceeded: true,
        reason: `Command limit exceeded: ${usage.totalCommands}/${FREE_TIER_LIMITS.commandsPerDay}`,
      };
    }

    if (usage.totalAgents >= FREE_TIER_LIMITS.agentsPerDay) {
      return {
        exceeded: true,
        reason: `Agent limit exceeded: ${usage.totalAgents}/${FREE_TIER_LIMITS.agentsPerDay}`,
      };
    }

    if (usage.totalPipelines >= FREE_TIER_LIMITS.pipelinesPerDay) {
      return {
        exceeded: true,
        reason: `Pipeline limit exceeded: ${usage.totalPipelines}/${FREE_TIER_LIMITS.pipelinesPerDay}`,
      };
    }

    return { exceeded: false };
  }

  /**
   * Get remaining free tier quota
   */
  getFreeTierRemaining(usage: DailyUsage): {
    commandsRemaining: number;
    agentsRemaining: number;
    pipelinesRemaining: number;
  } {
    return {
      commandsRemaining: Math.max(0, FREE_TIER_LIMITS.commandsPerDay - usage.totalCommands),
      agentsRemaining: Math.max(0, FREE_TIER_LIMITS.agentsPerDay - usage.totalAgents),
      pipelinesRemaining: Math.max(0, FREE_TIER_LIMITS.pipelinesPerDay - usage.totalPipelines),
    };
  }

  /**
   * Get usage report for CLI display
   */
  getUsageReport(licenseKey: string, days: number = 7): UsageReport {
    const today = new Date();
    const reports: DailyUsage[] = [];
    let totalCommands = 0;
    let totalAgents = 0;
    let totalPipelines = 0;

    for (let i = days - 1; i >= 0; i--) {
      const date = new Date(today);
      date.setDate(date.getDate() - i);
      const dateStr = date.toISOString().slice(0, 10);
      const dailyUsage = this.getDailyUsage(licenseKey, dateStr);
      reports.push(dailyUsage);
      totalCommands += dailyUsage.totalCommands;
      totalAgents += dailyUsage.totalAgents;
      totalPipelines += dailyUsage.totalPipelines;
    }

    return {
      licenseKeyHash: this.hashLicenseKey(licenseKey),
      periodDays: days,
      totalCommands,
      totalAgents,
      totalPipelines,
      dailyReports: reports,
    };
  }

  /**
   * Close database connection
   */
  close(): void {
    this.db.close();
  }
}

/**
 * Usage report structure
 */
export interface UsageReport {
  licenseKeyHash: string;
  periodDays: number;
  totalCommands: number;
  totalAgents: number;
  totalPipelines: number;
  dailyReports: DailyUsage[];
}

// Export singleton
export const usageTracker = UsageTracker.getInstance();

// Export convenience functions
export async function trackCommand(
  licenseKey: string,
  command: string,
  metadata?: Record<string, unknown>
): Promise<void> {
  await UsageTracker.getInstance().trackCommand(licenseKey, command, metadata);
}

export async function trackAgentCall(
  licenseKey: string,
  agentName: string,
  metadata?: Record<string, unknown>
): Promise<void> {
  await UsageTracker.getInstance().trackAgentCall(licenseKey, agentName, metadata);
}

export async function trackPipelineRun(
  licenseKey: string,
  pipelineType: string,
  metadata?: Record<string, unknown>
): Promise<void> {
  await UsageTracker.getInstance().trackPipelineRun(licenseKey, pipelineType, metadata);
}
